components/com_contact/src/Service/Router.php

```diff
@@ -241,20 +256,43 @@ public function getContactId($segment, $query)
             $dbquery = $this->db->getQuery(true);
             $dbquery->select($this->db->quoteName('id'))
                 ->from($this->db->quoteName('#__contact_details'))
-                ->where($this->db->quoteName('alias') . ' = :alias')
-                ->bind(':alias', $segment);
+                ->where($this->db->quoteName('alias') . ' = :segment')
+                ->bind(':segment', $segment);
 
             if (isset($query['id']) && $query['id']) {
-                $dbquery->where($this->db->quoteName('catid') . ' = :catid')
-                    ->bind(':catid', $query['id'], ParameterType::INTEGER);
+                $dbquery->where($this->db->quoteName('catid') . ' = :id')
+                    ->bind(':id', $query['id'], ParameterType::INTEGER);
             }
 
             $this->db->setQuery($dbquery);
 
-            return (int) $this->db->loadResult();
+            $id = (int) $this->db->loadResult();
+
+            // Do we have a URL with ID?
+            if ($id) {
+                return $id;
+            }
+
+            $this->app->getRouter()->setTainted();
+        }
+
+        $id = (int) $segment;
+
+        if ($id) {
+            $dbquery = $this->db->getQuery(true);
+            $dbquery->select($this->db->quoteName('alias'))
+                ->from($this->db->quoteName('#__contact_details'))
+                ->where($this->db->quoteName('id') . ' = :id')
+                ->bind(':id', $id);
```

Comment:
```suggestion
                ->bind(':id', $id, ParameterType::INTEGER);
```
Repeat with other instances.